portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -20,59 +20,128 @@
+    ret += "]";
+    return ret;
+  }
+
   private String getParameter(HttpServletRequest request, String name, String defaultValue) {
     String value = request.getParameter(name);
     if (value == null) {
       return defaultValue;
     }
     return value;
   }
-  // our datastructure for now.... very simple
-  private ArrayList<String> commentList = new ArrayList<String>();//Creating arraylist
+
+  private Query getQuerySort(String sort_method) {
+    if (sort_method.equals("ts_desc")) {
+        return new Query("Comment").addSort("timestamp", SortDirection.DESCENDING);
+    }
+    else if (sort_method.equals("ts_asc")) {
+        return new Query("Comment").addSort("timestamp", SortDirection.ASCENDING);
+    }
+    else if (sort_method.equals("rtg_desc")) {
+        return new Query("Comment").addSort("rating", SortDirection.DESCENDING);
+    }
+    else if (sort_method.equals("rtg_asc")) {
+        return new Query("Comment").addSort("rating", SortDirection.ASCENDING);
+    }
+    else if (sort_method.equals("author_desc")) {
+        return new Query("Comment").addSort("name", SortDirection.DESCENDING);
+    }
+    else if (sort_method.equals("author_asc")) {
+        return new Query("Comment").addSort("name", SortDirection.ASCENDING);
+    }
+    else if (sort_method.equals("comm_len_desc")) {
+        return new Query("Comment").addSort("comment_length", SortDirection.DESCENDING);
+    }
+    else if(sort_method=="comm_len_asc") {
+        return new Query("Comment").addSort("comment_length", SortDirection.ASCENDING);
+    }
+    return new Query("Comment").addSort("timestamp", SortDirection.DESCENDING);
+  }
+
+  private static double round (double value, int precision) {
+    int scale = (int) Math.pow(10, precision);
+    return (double) Math.round(value * scale) / scale;
+    }
+
   //get method
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    Query query = new Query("Comment").addSort("timestamp", SortDirection.DESCENDING);
+    String sort_method = (String) getParameter(request,"sort", "ts_desc");
+    Query query = getQuerySort(sort_method);
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
     PreparedQuery r = datastore.prepare(query);
+    List<Entity> results = r.asList(FetchOptions.Builder.withDefaults());
 
     int max_comments = Integer.parseInt(getParameter(request,"max-comments", "5"));
-    ArrayList<String> comments = new ArrayList<String>();
-
-    List<Entity> results = r.asList(FetchOptions.Builder.withDefaults());
+    ArrayList<Comment> comments = new ArrayList<Comment>();  
+    
     for (int i = 0; i < Math.min(max_comments, results.size()); i++) {
       Entity entity = results.get(i);
       String name = (String) entity.getProperty("name");
-      String comment = (String) entity.getProperty("comment");
-      //int rating = Integer.parseInt(entity.getProperty("rating"));
```

Comment:
I believe that in java, you can cast a long to and int, but you cannot cast a Long to any of {long, Integer, int}. Take a close look at the rules here:

https://docs.oracle.com/javase/tutorial/java/data/autoboxing.html

We can talk further in person.

Reply:
Resolved